### PR TITLE
bump protoc-gen-validate to latest

### PR DESCRIPTION
**Pull request by SecurityInsanity**
Description:

this commit bumps protoc-gen-validate to the latest version. this
should unblock `wrowe` in slack working on windows support.

after this I believe we can also take use of the new address validation
type to fix some unique error messages, but the first step is bumping it

as a side note:

- SocketState was using `.message.required` however it was not
  a message type. as far as I can tell this was a bug that PGV fixed.
  the proper annotation is .any.required

Signed-off-by: Cynthia Coan <ccoan@instructure.com>

Risk Level: Low
Testing: Ensure that envoy successfully builds.
Docs Changes: None
Release Notes: None

### CODE REVIEW

**Comment by SecurityInsanity** [comment]
Looks like tsan ran out of memory? At least that's the consensus on that error message looking through slack. Going to try out the fancy:

/retest

**Comment by SecurityInsanity** [comment]
Looks like there's another windows fix to PGV coming down the pipe now: https://github.com/envoyproxy/protoc-gen-validate/pull/244 I'm going to wait for that to be merged, and then bump this PR. Just so that way we don't have two seperate PRs.

**Comment by wrowe** [comment]
Things are better with the latest winsock2.h fix and regression fix for PGV, but I'm seeing several warnings and one windows failure I haven't unwound, and am unlike to be able to retest this today;

INFO: From GoLink external/com_envoyproxy_protoc_gen_validate/windows_amd64_stripped/protoc-gen-validate.exe [for host]:
link: warning: package "github.com/golang/protobuf/ptypes/any" is provided by more than one rule:
    @com_github_golang_protobuf//ptypes/any:go_default_library
    @io_bazel_rules_go//proto/wkt:any_go_proto
Set "importmap" to different paths in each library.
This will be an error in the future.
link: warning: package "github.com/golang/protobuf/ptypes/duration" is provided by more than one rule:
    @com_github_golang_protobuf//ptypes/duration:go_default_library
    @io_bazel_rules_go//proto/wkt:duration_go_proto
Set "importmap" to different paths in each library.
This will be an error in the future.
link: warning: package "github.com/golang/protobuf/ptypes/timestamp" is provided by more than one rule:
    @com_github_golang_protobuf//ptypes/timestamp:go_default_library
    @io_bazel_rules_go//proto/wkt:timestamp_go_proto
Set "importmap" to different paths in each library.
This will be an error in the future.
INFO: From CcCmakeMakeRule external/envoy/bazel/foreign_cc/ares/include:

ERROR: C:/_eb/external/envoy_api/envoy/api/v2/core/BUILD:34:1: C++ compilation of rule '@envoy_api//envoy/api/v2/core:base_cc' failed (Exit 2)
C:\_eb\execroot\envoy\bazel-out\x64_windows-dbg\bin\external\envoy_api\envoy\api\v2\core\base.pb.h(179): error C2059: syntax error: '('
C:\_eb\execroot\envoy\bazel-out\x64_windows-dbg\bin\external\envoy_api\envoy\api\v2\core\base.pb.h(179): error C3805: '(': unexpected token, expected either '}' or a ','
C:\_eb\execroot\envoy\bazel-out\x64_windows-dbg\bin\external\envoy_api\envoy\api\v2\core\base.pb.h(186): error C2143: syntax error: missing ')' before ';'
C:\_eb\execroot\envoy\bazel-out\x64_windows-dbg\bin\external\envoy_api\envoy\api\v2\core\base.pb.h(189): error C2065: 'PATCH': undeclared identifier

What is odd, here is what vim tells me about that file;

176   HEAD = 2,
177   POST = 3,
178   PUT = 4,
179   DELETE = 5,
180   CONNECT = 6,
181   OPTIONS = 7,
182   TRACE = 8,
183   PATCH = 9,
184   RequestMethod_INT_MIN_SENTINEL_DO_NOT_USE_ = std::numeric_limits<::google::protobuf::int32>::min(),
185   RequestMethod_INT_MAX_SENTINEL_DO_NOT_USE_ = std::numeric_limits<::google::protobuf::int32>::max()
186 };
187 bool RequestMethod_IsValid(int value);
188 const RequestMethod RequestMethod_MIN = METHOD_UNSPECIFIED;
189 const RequestMethod RequestMethod_MAX = PATCH;
190 const int RequestMethod_ARRAYSIZE = RequestMethod_MAX + 1;
191
192 const ::google::protobuf::EnumDescriptor* RequestMethod_descriptor();    
```

**Comment by wrowe** [comment]
The attached file corresponds to the bazel-out\x64_windows-dbg\bin\external\envoy_api\envoy\api\v2\core\base.pb.h file which is failing.

There is a significant clue buried in 

./bazel-bin/source/common/filesystem/_virtual_includes/directory_iterator_impl_lib_win32/common/filesystem/directory_iterator_impl.h
// <windows.h> uses macros to #define a ton of symbols, two of which (DELETE and GetMessage)
...

I'd pursue but am a bit worn out this afternoon, so my next chance to retest will be sometime over the weekend or Monday. Cheers and thanks for all the help to resolve these quirks!

**Comment by SecurityInsanity** [comment]
Sounds good. I’ll pull in the latest update after lunch, and I think we should merge this PR. We can then handle the windows errors next week in more PRs.

File: api/envoy/api/v2/core/base.proto [hunk 1]
```diff
@@ -259,7 +259,7 @@ message SocketOption {
   // The state in which the option will be applied. When used in BindConfig
   // STATE_PREBIND is currently the only valid value.
   SocketState state = 6
-      [(validate.rules).message.required = true, (validate.rules).enum.defined_only = true];
+      [(validate.rules).any.required = true, (validate.rules).enum.defined_only = true];
```
Comment (lizan) [inline api/envoy/api/v2/core/base.proto]: hmm, remove `(validate.rules).any.required = true`? SocketState is an enum and that should do nothing.

Comment (SecurityInsanity) [reply api/envoy/api/v2/core/base.proto]: I think that’s still required to ensure the field is not, not provided.

Any required ensure a value is provided. And then enum defined_only ensures it’s not bogus

Comment (lizan) [reply api/envoy/api/v2/core/base.proto]: No, proto3 doesn't, and have no way to, differentiate non-set vs default-value for non-message types so it is no-op. https://developers.google.com/protocol-buffers/docs/proto3#default

Comment (SecurityInsanity) [reply api/envoy/api/v2/core/base.proto]: Ah you're right my bad. I'll push up another commit that removes it. Might be worth opening up a ticket to find other bad uses of required.

File: api/bazel/repository_locations.bzl [hunk 2]
```diff
@@ -7,8 +7,8 @@ GOGOPROTO_SHA256 = "99e423905ba8921e86817607a5294ffeedb66fdd4a85efce5eb2848f715f
 OPENCENSUS_PROTO_GIT_SHA = "d5d80953a8c2ff4633087af6933cd152678434bb"  # Mar 18, 2019
 OPENCENSUS_PROTO_SHA256 = "a4e87a1da21d1b3a16674332c3ee6e2689d52f3532e2ff8cb4a626c8bcdabcfc"
 
-PGV_GIT_SHA = "26db5cb7c01a67c6a2e21a832106406185530b2f"
-PGV_SHA256 = "6510cbcf69d99059c652ae2376f6240bc761d0b019cd962225f4f609be361e26"
+PGV_GIT_SHA = "dd90514d96e35023ed20201f349542b0bc64461d"
```
Comment (moderation) [inline api/bazel/repository_locations.bzl]: @SecurityInsanity Would have been nice to refer to the commit date of July 19 2019 like we do with other non-release dependencies. As a TODO we should switch to PGV release 1.1 when it is released and assuming it contains this commit. Sorry I didn't pick this up earlier.

Comment (SecurityInsanity) [reply api/bazel/repository_locations.bzl]: Hey @moderation , yeah we should switch to more stabilized versions, and comment for sure. However, keeping status quo here I don't think is particularly bad. 

Something to remember, but not all that important for now.

**Comment by achasveachas** [comment]
> The attached file corresponds to the bazel-out\x64_windows-dbg\bin\external\envoy_api\envoy\api\v2\core\base.pb.h file which is failing.
> 
> There is a significant clue buried in
> 
> ./bazel-bin/source/common/filesystem/_virtual_includes/directory_iterator_impl_lib_win32/common/filesystem/directory_iterator_impl.h
> // <windows.h> uses macros to #define a ton of symbols, two of which (DELETE and GetMessage)
> ...
> 
> I'd pursue but am a bit worn out this afternoon, so my next chance to retest will be sometime over the weekend or Monday. Cheers and thanks for all the help to resolve these quirks!
> 
> [
> [base.pb.h.txt](https://github.com/envoyproxy/envoy/files/3412222/base.pb.h.txt)
> ](url)

This seems to be related to the issue referenced inthis comment:

https://github.com/envoyproxy/envoy/blob/b09a6d4a33137c5547fb6aec03bdaa07214b260f/source/common/filesystem/win32/directory_iterator_impl.h#L5

We probably just need to use the same workaround.

**Comment by achasveachas** [comment]
Submitted https://github.com/envoyproxy/protoc-gen-validate/pull/247

**Pull request merged**